CMakeLists.txt

```diff
@@ -49,6 +49,28 @@ function(packageProject)
     endif()
 
     if(DEFINED PROJECT_VERSION_HEADER)
+      string(REPLACE "." ";" VERSION_LIST "${PROJECT_VERSION}")
```

Comment:
Imo it could make sense to accept any valid semantic version, in which case we might want to use the [official regex](https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string) instead.

Reply:
CMake doesn't support other semantic versions:
@<https://cmake.org/cmake/help/latest/command/project.html#command:project>
> VERSION <version>
Takes a <version> argument composed of non-negative integer components, i.e. <major>[.<minor>[.<patch>[.<tweak>]]], and sets the variables

Reply:
I see your point. `PackageProject` however allows passing arbitrary version strings through the `VERSION` argument. I see some cases where users would like to add metadata (e.g. git commit hash) to their version string. Currently we're also not enforcing the integer requirement, so such versions would create C++ syntax errors, which would be very unexpected to some users.

Using the regex with CMake's [regex match](https://cmake.org/cmake/help/latest/command/string.html#regex-match) should allow more general versions and also avoid such mistakes.

Reply:
Hmm, I see what you mean, but what would the `*_MAJOR` etc. versions give in that case?

Reply:
Not sure I understand your question, but I believe you'd call something like

```
string(REGEX MATCH "^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$" match "${PROJECT_VERSION}")
```

And then the `MAJOR`, `MINOR` versions would be stored in `CMAKE_MATCH_<n>` variables.

> And one with numbered capture groups instead (so cg1 = major, cg2 = minor, cg3 = patch, cg4 = prerelease and cg5 = buildmetadata) that is compatible with ECMA Script (JavaScript), PCRE (Perl Compatible Regular Expressions, i.e. Perl, PHP and R), Python and Go.

Reply:
I meant, if `1.0.4.testing` is passed to `VERSION` then `Project_VERSION_BUILD` would be:
```cpp
#define PROJECT_VERSION_BUILD testing
```

In that case, wouldn't it be better if it was stringified? Or how do we make it so that the macro holds a sensible value?

Reply:
Oh yeah from a quick look, the regex seems to guarantee only that `major`, `minor` and `patch` to be integer. In that case I'd stringify the `prerelease` and `buildmetadata` groups, but I think it would still be nice to be able to use integers for the guaranteed numeric components.

Reply:
Yeah I agree, but how do we make it stringified if it's not an integer?

Reply:
We could just quote the argument in the header, as they are not allowed from the regex afaict:

```cpp
#define @UPPERCASE_PROJECT_NAME@_VERSION_PRERELEASE "@PROJECT_VERSION_PRERELEASE@"
#define @UPPERCASE_PROJECT_NAME@_VERSION_BUILDMETADATA "@PROJECT_VERSION_BUILDMETADATA@"
```

Reply:
Those two variables don't exist at the moment. I was asking about the `*_BUILD` macro and how to switch it from integer to string.

Reply:
Yeah I used the according names from the semantic versioning definitions. In any case, as long as they don't contain any `"` characters, simply quoting the variable in the header should work?

Reply:
Oh, I misunderstood your post from before.

> Oh yeah from a quick look, the regex seems to guarantee only that major, minor and patch to be integer. In that case I'd stringify the prerelease and buildmetadata groups, but I think it would still be nice to be able to use integers for the **guaranteed numeric components**.

I read that as if you wanted the build category to be integer if it's an integer and string if it's a string.

I get it now, I'll modify the PR.

Reply:
I'm mot sure we understood each other, in the current implementation there is no checking if the arguments are indeed in the correct Format. E.g. if we would pass a typical semantic version string such as ` 1.2.3-a` to `PackageProject`, the build will fail or induce unexpected behaviour. Using a regex to ensure that a version is indeed in the proper format would be a solution to this problem.

I'm aware that CMake uses a different versioning scheme `maj.min.patch.tweak`, instead of the widely adopted semantic versioning. Imo we can keep the CMake style versioning, but at least check the version arg for consistency to avoid problems with compilation. 

To be precise, how about instead of `string(REPLACE "." ";" VERSION_LIST "${PROJECT_VERSION}")` to separate the version string, we use something like

```cmake
string(REGEX MATCH "^(0|[1-9]\\d*)(\\.(0|[1-9]\\d*))?(\\.(0|[1-9]\\d*))?(\\.(0|[1-9]\\d*))?$" _ "${PROJECT_VERSION}")
```

and read the version components from `CMAKE_MATCH_<N>`. Then we can be sure that all arguments are either integer or empty strings.

Reply:
I can't get the regex to work. I've tried a thousand things, none work. Can you try it yourself, I really have no idea what I'm doing wrong. I've also simplified the regex to this: `^([0-9]+)(\\.([0-9]+))?(\\.([0-9]+))?(\\.([0-9]+))?$`

The full thing I have is just this, it doesn't even populate CMAKE_MATCH_<N>:

```cmake
string(REGEX MATCH "^([0-9]+)(\\.([0-9]+))?(\\.([0-9]+))?(\\.([0-9]+))?$" _ "${PROJECT_VERSION}")

set(PROJECT_VERSION_MAJOR ${REGEX_MATCH_1})
set(PROJECT_VERSION_MINOR ${REGEX_MATCH_3})
set(PROJECT_VERSION_PATCH ${REGEX_MATCH_5})
set(PROJECT_VERSION_TWEAK ${REGEX_MATCH_7})

if(NOT DEFINED PROJECT_VERSION_MAJOR)
  set(PROJECT_VERSION_MAJOR "0")
endif()
if(NOT DEFINED PROJECT_VERSION_MINOR)
  set(PROJECT_VERSION_MINOR "0")
endif()
if(NOT DEFINED PROJECT_VERSION_PATCH)
  set(PROJECT_VERSION_PATCH "0")
endif()
if(NOT DEFINED PROJECT_VERSION_TWEAK)
  set(PROJECT_VERSION_TWEAK "0")
endif()
```

Reply:
Shouldn't it be `CMAKE_MATCH_<N>` instead of `REGEX_MATCH_<N>`?

Reply:
I'm blind, I apologize lol...